### ISSUE TITLE
How to compute the two-order partial derivative in a non-graph based framework

### ISSUE THREAD

**Issue by zsffq999** [igul222/improved_wgan_training#4]
I have noticed that this work is implemented by tensorflow where the graph of the gradient can be constructed. I wonder how to compute two-order partial derivative with non-graph based deep-learning framework like torch/pytorch/etc. It seems impossible to optimize the norm of the gradient with these frameworks.

Anyway, computing the gradient of norm of the gradient involves the dot product of Jacobian matrix and the gradient, thus the computation may be expensive. I wonder the efficiency of improved-wgan in computing the gradient?

**Comment by igul222** [comment]
It's not currently possible to implement this model in torch/pytorch unfortunately, but there's an open PR which will add support for gradients-of-gradients, at which point it might be possible: https://github.com/pytorch/pytorch/pull/1016

Computing gradients of gradients is typically done in tensorflow/theano/etc. as backprop through backprop, with 2x compute cost of just computing the gradients. It's done in several other works, https://arxiv.org/abs/1611.02163 and https://arxiv.org/abs/1502.03492 are the first that come to mind.

**Comment by apaszke** [comment]
@igul222 actually it is possible, but you'll need to compute the backward twice, and estimate the hessian-vector product using a finite difference approximation. See [this tweet](https://twitter.com/soumithchintala/status/850509545524822016) for more details.

**Comment by qianguih** [comment]
@apaszke Thanks for the note! However, I'm sorry that I'm still confused. Is there any example showing how to do this in pytorch?

**Comment by apaszke** [comment]
Unfortunately I'm not aware of any examples.
